src/main/java/com/mixpanel/android/mpmetrics/GCMReceiver.java

```diff
@@ -96,6 +101,118 @@ public void onReceive(Context context, Intent intent) {
         }
     }
 
+    /*
+     * Package scope for testing only, do not call outside of GCMReceiver.
+     */
+
+    /* package */ static class NotificationData {
+        private NotificationData(int anIcon, CharSequence aTitle, String aMessage, Intent anIntent) {
+            icon = anIcon;
+            title = aTitle;
+            message = aMessage;
+            intent = anIntent;
+        }
+
+        public final int icon;
+        public final CharSequence title;
+        public final String message;
+        public final Intent intent;
+    }
+
+    /* package */ Intent getDefaultIntent(Context context) {
+        final PackageManager manager = context.getPackageManager();
+        return manager.getLaunchIntentForPackage(context.getPackageName());
+    }
+
+    /* package */ NotificationData readInboundIntent(Context context, Intent inboundIntent, ResourceIds iconIds) {
+        final PackageManager manager = context.getPackageManager();
+
+        final String message = inboundIntent.getStringExtra("mp_message");
+        final String iconName = inboundIntent.getStringExtra("mp_icnm");
+        final String uriString = inboundIntent.getStringExtra("mp_cta");
+        CharSequence notificationTitle = inboundIntent.getStringExtra("mp_title");
+
+        if (message == null) {
+            return null;
+        }
+
+        int notificationIcon = -1;
+        if (null != iconName) {
+            if (iconIds.knownIdName(iconName)) {
+                notificationIcon = iconIds.idFromName(iconName);
+            }
+        }
+
+        ApplicationInfo appInfo;
+        try {
+            appInfo = manager.getApplicationInfo(context.getPackageName(), 0);
+        } catch (final NameNotFoundException e) {
+            appInfo = null;
+        }
+
+        if (notificationIcon == -1 && null != appInfo) {
+            notificationIcon = appInfo.icon;
+        }
+
+        if (notificationIcon == -1) {
+            notificationIcon = android.R.drawable.sym_def_app_icon;
+        }
+
+        if (null == notificationTitle && null != appInfo) {
+            notificationTitle = manager.getApplicationLabel(appInfo);
+        }
+
+        if (null == notificationTitle) {
+            notificationTitle = "A Message For You";
```

Comment:
In standard Mixpanel form we should probably not Title Case this (even though it's technically a title)